File: packages/enterprise/src/audit.ts
/**
 * Audit logging: structured events for LLM calls, tool invocations, memory access.
 * Scalable: async, pluggable sinks; durable: persistence; secure: no secrets in logs.
 */
export type AuditEventType =
  | "llm:chat"
  | "llm:stream"
  | "tool:execute"
  | "memory:read"
  | "memory:write"
  | "memory:search"
  | "delegation"
  | "hitl:intercept"
  | "policy:deny";

export interface AuditEvent {
  type: AuditEventType;
  tenantId?: string;
  userId?: string;
  threadId?: string;
  timestamp: number;
  metadata?: Record<string, unknown>;
  /** Redacted; never log raw secrets */
  resource?: string;
}

export interface AuditLogger {
  log(event: AuditEvent): void | Promise<void>;
}

File: python/src/agent_framework/enterprise/audit.py
"""Audit logging."""

from abc import ABC, abstractmethod
from typing import Literal, Any

AuditEventType = Literal[
    "llm:chat", "llm:stream", "tool:execute",
    "memory:read", "memory:write", "memory:search",
    "delegation", "hitl:intercept", "policy:deny",
]


class AuditEvent:
    def __init__(
        self,
        type: str,
        timestamp: float,
        tenant_id: str | None = None,
        user_id: str | None = None,
        thread_id: str | None = None,
        metadata: dict[str, Any] | None = None,
        resource: str | None = None,
    ):
        self.type = type
        self.timestamp = timestamp
        self.tenant_id = tenant_id
        self.user_id = user_id
        self.thread_id = thread_id
        self.metadata = metadata or {}
        self.resource = resource


class AuditLogger(ABC):
    @abstractmethod
    async def log(self, event: AuditEvent) -> None:
        ...
